File: app/src/main/java/in/juspay/devtools/CheckoutActivity.java
package in.juspay.devtools;

import androidx.appcompat.app.AppCompatActivity;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.ProgressBar;
import in.juspay.hypercheckoutlite.HyperCheckoutLite;
import in.juspay.hypersdk.data.JuspayResponseHandler;
import in.juspay.hypersdk.ui.HyperPaymentsCallbackAdapter;
import org.json.JSONException;
import org.json.JSONObject;
import java.io.IOException;

public class CheckoutActivity extends AppCompatActivity {
    private Button processButton;
    private CoordinatorLayout coordinatorLayout;
    private ProgressBar progressBar;
    private ImageView backImage;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_checkout);
    }

    @Override
    protected void onStart() {
        super.onStart();
        updatingUI(); //This is relevant only for sample project
        processButton = findViewById(R.id.rectangle_9);
        processButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                runOnUiThread(() -> {
                    progressBar.setVisibility(View.VISIBLE);
                });
                try {
                    startPayment();
                } catch (Exception e) {

                }
            }
        });
        backImage = findViewById(R.id.imageView);
        backImage.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                onBackPressed();
            }
        });
    }

    // block:start:startPayment
    private void startPayment() throws IOException {
        JSONObject payload = new JSONObject();

        // block:start:sendPostRequest
        ApiClient.sendPostRequest("http://10.0.2.2:5000/initiateJuspayPayment", payload, new ApiClient.ApiResponseCallback() {
            @Override
            public void onResponseReceived(String response) throws JSONException {
                try {
                    JSONObject sdkPayload = new JSONObject(response).getJSONObject("sdkPayload");
                    CheckoutActivity.this.runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            // block:start:openPaymentPage
                            HyperCheckoutLite.openPaymentPage(CheckoutActivity.this, sdkPayload, createHyperPaymentsCallbackAdapter());
                            // block:end:openPaymentPage
                            new Helper().showSnackbar("Opening Payment Page", coordinatorLayout);
                        }
                    });
                } catch (Exception e) {
                    Log.d("EXCEPTATION:", e.toString());
                    new Helper().showSnackbar("Unable to parse sdkPayload from response", coordinatorLayout);
                }
            }

            @Override
            public void onFailure(Exception e) {
                Log.d("EXCEPTATION: ", e.toString());
                new Helper().showSnackbar("Payment Initiation API Failed", coordinatorLayout);
            }
        });
        // block:end:sendPostRequest
    }
    // block:end:startPayment

    @Override
    protected void onPause() {
        super.onPause();
    }

    // block:start:create-hyper-callback
    private HyperPaymentsCallbackAdapter createHyperPaymentsCallbackAdapter() {
        return new HyperPaymentsCallbackAdapter() {
            @Override
            public void onEvent(JSONObject jsonObject, JuspayResponseHandler responseHandler) {
                Intent redirect = new Intent(CheckoutActivity.this, ResponsePage.class);
                try {
                    String event = jsonObject.getString("event");
                    if (event.equals("hide_loader")) {
                        // Hide Loader
                        runOnUiThread(() -> {
                            progressBar.setVisibility(View.GONE);
                        });
                    } else if (event.equals("process_result")) {
                        JSONObject innerPayload = jsonObject.optJSONObject("payload");
                        String status = innerPayload.optString("status");
                        switch (status) {
                            case "backpressed":
                            case "user_aborted":
                                // handle back

                                new Helper().showSnackbar("User Aborted", coordinatorLayout);
                                break;
                            default:
                                String orderId = jsonObject.getString("orderId");
                                redirect.putExtra("orderId", orderId);
                                startActivity(redirect);
                        }
                    }
                    // block:end:handle-process-result
                } catch (Exception e) {
                    // merchant code...
                }
            }
        };
    }
    // block:end:create-hyper-callback


    //    block:start:onBackPressed
    @Override
    public void onBackPressed() {
        if (!HyperCheckoutLite.onBackPressed()) {
            // Handle Backpress Here
            super.onBackPressed();
        }
    }
    //block:end:onBackPressed


    private void updatingUI() {
        coordinatorLayout = findViewById(R.id.coordinatorLayout);
        progressBar = findViewById(R.id.progressBar);
    }
}

File: app/src/main/java/in/juspay/devtools/CheckoutActivity.kt
package `in`.juspay.devtools

import android.app.ProgressDialog
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.widget.Button
import androidx.appcompat.app.AppCompatActivity
import androidx.coordinatorlayout.widget.CoordinatorLayout
import `in`.juspay.hypercheckoutlite.HyperCheckoutLite
import `in`.juspay.hypersdk.data.JuspayResponseHandler
import `in`.juspay.hypersdk.ui.HyperPaymentsCallbackAdapter
import org.json.JSONObject

class CheckoutActivity : AppCompatActivity() {
    private lateinit var proceedButton: Button
    private lateinit var dialog: ProgressDialog
    private lateinit var coordinatorLayout: CoordinatorLayout
    private var orderId: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_products)
    }

    override fun onStart() {
        super.onStart()
        coordinatorLayout = findViewById(R.id.coordinatorLayout2)
        dialog = ProgressDialog(this@CheckoutActivity)
        dialog.setMessage("Processing...")
        proceedButton = findViewById(R.id.rectangle_8)
        proceedButton.setOnClickListener {
            dialog.show()
            try {
                startPayment()
            } catch (e: Exception) {
            }
        }
    }

    override fun onBackPressed() {
        val handleBackpress = HyperCheckoutLite.onBackPressed()
        if (!handleBackpress) {
            super.onBackPressed()
        }
    }

    private fun startPayment() {
        var payload = JSONObject();
        var itemDetail1 = JSONObject();
        itemDetail1.put("item_id", "12345");
        itemDetail1.put("quantity", 2);

        var itemDetail2 = JSONObject();
        itemDetail2.put("item_id", "54321");
        itemDetail2.put("quantity", 2);

        val itemDetails = arrayOf(itemDetail1, itemDetail2);
        payload.put("item_details", itemDetails);


        ApiClient.sendPostRequest("http://10.0.2.2:5000/initiateJuspayPayment", payload, object : ApiClient.ApiResponseCallback {
            override fun onResponseReceived(response: String?) {
                try {
                    val sdkPayload = JSONObject(response).getJSONObject("sdkPayload")
                    try {
                        orderId = JSONObject(response).getString("orderId")
                    }catch (e: Exception) {
                        Log.d("ERROR>>>", e.toString())
                    }
                    orderId = JSONObject(response).getString("orderId")
                    runOnUiThread {
                        HyperCheckoutLite.openPaymentPage(this@CheckoutActivity, sdkPayload, createHyperPaymentsCallbackAdapter())
                    }
                } catch (e: Exception) {
                    Log.d("ERROR>>>", e.toString())
                }
            }

            override fun onFailure(e: Exception?) {

            }
        })
    }

    private fun createHyperPaymentsCallbackAdapter(): HyperPaymentsCallbackAdapter {
        return object : HyperPaymentsCallbackAdapter() {
            override fun onEvent(jsonObject: JSONObject, responseHandler: JuspayResponseHandler) {
                val redirect = Intent(this@CheckoutActivity, ResponsePage::class.java)
                try {
                    val event = jsonObject.getString("event")
                    when (event) {
                        "hide_loader" -> {
                            // Hide Loader
                            dialog.hide()
                        }
                        "process_result" -> {
                            val innerPayload = jsonObject.optJSONObject("payload")
                            val status = innerPayload?.optString("status")
                            when (status) {
                                "backpressed", "user_aborted" -> {
                                    redirect.putExtra("orderId", orderId)
                                    startActivity(redirect)
                                }
                                else -> {
                                    redirect.putExtra("orderId", orderId)
                                    startActivity(redirect)
                                }
                            }
                        }
                    }
                } catch (e: Exception) {
                    // merchant code...
                }
            }
        }
    }
}
